projects/RabbitMQ.Client.OAuth2/OAuth2CredentialsProvider.cs

```diff
@@ -31,112 +31,81 @@
 
 using System;
 using System.Threading;
+using System.Threading.Tasks;
 
 namespace RabbitMQ.Client.OAuth2
 {
     public class OAuth2ClientCredentialsProvider : ICredentialsProvider
     {
-        const int TOKEN_RETRIEVAL_TIMEOUT = 5000;
-        private ReaderWriterLock _lock = new ReaderWriterLock();
+        // TODO Dispose
+        private readonly ReaderWriterLockSlim _lock = new ReaderWriterLockSlim();
```

Comment:
I don't think ReadWriterLockSlim is valid for Async/await, you need to go the SemaphoreSlim route.